### PR TITLE
ch4: Collective Error Handling

### PR DESCRIPTION
**Pull request by wesbland**
Fix collective error handling in CH4 by adding new functions to the ADI for send/receive operations for collectives.

This follows the same scheme as the CH3 model of collective error handling where collectives will complete all of the matching even if there is an error and the data is corrupt/missing.

The PR also adds more tests to try different collectives with errors.

See #3775

### CODE REVIEW

**Comment by raffenet** [comment]
Will this also fix the issues with the `bcastlength` test? If so, can you add a commit to remove it from `xfail.conf`?

**Comment by wesbland** [comment]
Yep. Good point. 👍

**Comment by wesbland** [comment]
I don't think there's a specific issue open for this that I should be closing. Right?

**Comment by raffenet** [comment]
No, not that I'm aware of.

**Comment by hajimefu** [comment]
test:jenkins/ch3/tcp

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by wesbland** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by wesbland** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by wesbland** [comment]
test:jenkins/ch4/ucx

**Comment by wesbland** [comment]
test:jenkins/ch4/ucx

**Comment by wesbland** [comment]
@yfguo / @hzhou - This should be ready for someone to review now.

**Comment by wesbland** [comment]
test:jenkins/ch4/ucx

**Comment by wesbland** [comment]
I have no idea if the UCX xfails are up to date or not. There's some `gather_big` failures here. Are those normal or my fault?

**Comment by hzhou** [comment]
> 
> 
> I have no idea if the UCX xfails are up to date or not. There's some `gather_big` failures here. Are those normal or my fault?

Not your fault.

**Comment by hzhou** [comment]
Also, ignore the `new-quick` test. I am playing with the new Jenkins server and forgot to set the trigger properly.

**Comment by wesbland** [comment]
Sounds good. I'll leave this for review then.

File: src/mpid/ch4/generic/mpidig_send.h [hunk 1]
```diff
@@ -171,6 +233,34 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_send(const void *buf,
     return mpi_errno;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIDIG_mpi_send_coll
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_send_coll(const void *buf,
```
Comment (raffenet) [inline src/mpid/ch4/generic/mpidig_send.h]: This API should not have `_mpi_` in the name. The intention is for ch4 APIs with `_mpi_` to map directly to MPI functions.

File: src/mpid/ch4/netmod/ofi/ofi_events.h [hunk 2]
```diff
@@ -80,7 +80,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_recv_event(struct fi_cq_tagged_entry *wc,
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_RECV_EVENT);
 
     rreq->status.MPI_SOURCE = MPIDI_OFI_cqe_get_source(wc, true);
-    rreq->status.MPI_ERROR = MPI_SUCCESS;
+    MPIDI_OFI_idata_get_error_bits(wc->data, &rreq->status.MPI_ERROR);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_events.h]: I'd like to see if we can make this API look more like `MPIDI_OFI_cqe_get_source` above it. I.e. make error the return value. Maybe the prefix could also be `MPIDI_OFI_cqe_get_`?

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 3]
```diff
@@ -16,7 +16,8 @@
 
 MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send_lightweight(const void *buf,
                                                         size_t data_sz,
-                                                        int rank,
+                                                        int src_rank,
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_send.h]: Is this API change related necessary to this commit?

File: src/mpid/ch4/src/ch4_coll_impl.h [hunk 4]
```diff
@@ -21,6 +21,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Barrier_intra_composition_alpha(MPIR_Comm * c
                                                                    * ch4_algo_parameters_container)
 {
     int mpi_errno = MPI_SUCCESS;
+    int mpi_errno_ret = MPI_SUCCESS;
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_coll_impl.h]: Not a big fan of the naming here. Maybe `coll_err` or something else?

File: src/mpid/ch4/generic/mpidig_send.h (hunk 1 above)
Comment (wesbland) [reply src/mpid/ch4/generic/mpidig_send.h]: Yep. Good point. I'll take this out.

I'll also drop the `FUNCNAME` definition.

File: src/mpid/ch4/netmod/ofi/ofi_send.h (hunk 3 above)
Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_send.h]: Agreed. I've moved these out to another commit.

File: src/mpid/ch4/src/ch4_coll_impl.h (hunk 4 above)
Comment (wesbland) [reply src/mpid/ch4/src/ch4_coll_impl.h]: Agreed. I've renamed this.

File: src/mpid/ch4/netmod/ofi/ofi_events.h (hunk 2 above)
Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_events.h]: 👍 Done

**Comment by wesbland** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by wesbland** [comment]
@raffenet I think all your concerns have been addressed now.

File: src/mpid/ch4/netmod/ofi/ofi_send.h (hunk 3 above)
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_send.h]: This is adding `src_rank` to the argument list for send operations where it can be gotten from `comm->rank`. I think renaming to `dst_rank` is fine. The extra argument I'm not seeing the need for strongly.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_send.h]: I see what you mean now. I've taken this out.

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 5]
```diff
@@ -64,12 +64,12 @@ static inline void MPIDI_OFI_idata_set_error_bits(uint64_t * data_field, MPIR_Er
 }
 
 /* Get the error flag from the OFI data field. */
-static inline void MPIDI_OFI_idata_get_error_bits(uint64_t idata, int *err_res)
+static inline int MPIDI_OFI_idata_get_error_bits(uint64_t idata)
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_types.h]: Maybe I should have pointed this out on the commit that added these APIs. Because now we are adding them then immediately changing them when using.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: 🙄 Fixed. Sorry.
